api/v1/relay/relay.proto

```diff
@@ -0,0 +1,152 @@
+// Copyright 2020 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+syntax = "proto3";
+
+import "google/protobuf/timestamp.proto";
+import public "flow/flow.proto";
+
+package relay;
+
+service HubbleRelay {
+    rpc GetEvents(GetEventsRequest) returns (stream GetEventsResponse) {}
+}
+
+// Here I didn't include "follow", "until", and "number". This request assumes follow,
+// and lets the client decide when to end the request, whether it's based on timestamp
+// or the number of responses received.
+message GetEventsRequest{
+    // EventType specifies which types of events to subscribe to. If unspecified, the
+    // request is subscribed to all the event types.
+    repeated EventType event_types = 1;
+    repeated EventFilter blacklist = 2;
+    repeated EventFilter whitelist = 3;
+    google.protobuf.Timestamp since = 4;
+}
+
+message GetEventsResponse{
+    string node = 1;
+    google.protobuf.Timestamp timestamp = 2;
+    oneof event {
+        flow.Flow flow = 3;
+        K8sNamespaceState k8s_namespace_state = 4;
+        ServiceState service_state = 5;
+        // It's unclear how we can define what it means for service links to get
+        // added/deleted. For now I'm deferring the decision for whether a particular
+        // link should be included in the service map to the client. Maybe the server
+        // can periodically send service_link events for active links to keep-alive,
+        // and the client can expire older links based on their timestamp.
+        ServiceLinkState  service_link_state = 6;
+    }
+}
+
+// Here I added different filter types for different events. Another option is to reuse
+// and extend FlowFilter to filter all the events, but I worry that it might become too
+// confusing in terms of figuring out which filter fields apply to which events.
+message EventFilter {
+    oneof filter {
```

Comment:
I'm trying to understand the implications of this `oneof` here. If we treat the whitelist/blacklist the same as we do in the `service Observer`, then we will OR the `EventFilters` together, i.e. for an event to be included in the result, only a single filter has to match.

So if we want to return services and flows for namespace "foo", we have to do the following:
```
GetEventsRequest{
  event_types = [FLOW,SERVICE_STATE];
  whitelist = [{flow_filter={namespace="foo"}},{service_filter{namespace="foo"}}]
}
```

And if we have an event filter as follows:
```
GetEventsRequest{
  event_types = [FLOW,SERVICE_STATE];;
  whitelist = [{flow_filter={namespace="foo"}}];
}
```
Then this has to mean that we will only return flows from namespace "foo". No services will be returned, since a service cannot be matched by the flow filter. Is that a correct interpretation?

Maybe it would make sense to have a an additional `MetadataFilter` or something in here as well, which will be applied to each kind of event, so we can e.g. easily filter by namespace (which I think will be a common request):

```
GetEventsRequest{
  event_types = [FLOW,SERVICE_STATE];;
  whitelist = [{metadata_filter={namespace="foo"}}];
}
```
Would return both flows and services for the "foo" namespace.